### ISSUE TITLE
Optional Swift Protocol Food-Consumer Example Seems Inelegant

### ISSUE THREAD

**Issue by rileytestut** [ashfurrow/blog#46]
While your solution for your example of Swift's inability to have optional methods in non-Objc protocols is certainly clever, to me it seems to blur the responsibilities and roles of each component (of course, this is just my opinion, so take everything I say with a Swift grain of salt).

In your example, I view the Food objects as nothing more than pure model objects: they have properties, but don't really "do" anything themselves; their sole purpose is to provide the Consumer with information so the Consumer can then perform the needed functionality. However, by placing the responsibility on the Food objects to update the state of the Consumer, it blurs the lines between the two: is Consumer a pure model object? Is Food? Or are none of them pure model objects? Further more, what if you decided later to change the properties of the Consumer? You would then have to reflect this in all your Food implementations, which as we all know should be a warning sign that there should be some refactoring involved. Plus, this can then be seen as a tight-coupling of sorts; sure, you're using protocols, but if you have to change your implementation of how the Food objects "get eaten" every time you change the Consumer protocol, you lose a lot of the benefits of loose-coupling in the first place.

This, of course, is only an issue to me because it doesn't reflect my viewpoint of how things work in the "real world", which I try my best to represent in my code, since it makes the entire structure "feel" natural (but that is entirely up for debate). Should the components be changed to something else like a standardized test answer sheet and an answer sheet reader, to me it seems even more awkward. An answer sheet contains nothing but answers filled in by students; it doesn't do anything since it's essentially pure data. It should be entirely the responsibility of the answer sheet reader to a) read the input answer sheet, b) convert that into usable information, and then c) update its state (and the student's grade). Using your example would instead give the responsibility of updating the student's grade to the answer sheet itself, which breaks my preconceived notions of what is "smart" in this situation and what is "dumb".

Since I've spent a bit of time arguing why your solution feels "wrong", I should probably put in my own $0.02 over what I would do. Personally, I would just use the same solution you came to in your post, multiple protocols: a protocol for caloriesPerGram, and a protocol for milliLitersWaterPerGram. (Actually, my first choice would be to make both caloriesPerGram and milliLitersWaterPerGram return an optional value and have all Food instances implement them, since it also allows for food that doesn't have calories or water, but you didn't seem to like that idea).

tl;dr: I agree with your solution of using multiple protocols to solve Swift's lack of optional protocol methods, but I don't know why you didn't use the same solution for your Consumer-Food example. The solution you ended up using seemed more inelegant to me, because it mixed responsibilities of objects involved (resulting in a somewhat tight-coupling after all, to me at least), as well as didn't fit in with my view of how the world works (Food should be pure data IMO). However, this is all opinion, and the truth is the actual code is in fact clever and works, so don't let one person's opinion over it deter you from implementing something similar in production :D

**Comment by ashfurrow** [comment]
Hi @rileytestut – great, in-depth feedback. I appreciate it. The solution I proposed isn't optimal, for the reasons you've given above. I tried in the post to make it clear that this is _a_ solution, and that the best technique to solve a problem will vary as the problem does. The post wasn't so much about trying to get readers to do things my way, as it was trying to encourage them to look for alternate solutions to familiar problems. 

@owensd has written a great alternate approach using enums that you should check out: http://owensd.io/2015/02/02/swift-and-protocols.html
